### PR TITLE
KvStore - A data-store to support BlockTree

### LINKED ISSUE #2566: Replace DbLedger with simple blob store
## Problem

`DbLedger` currently requires `rocksdb`, while we use very little of what it has to offer. RocksDb types are also This also has significant build costs and is a large non-Rust dependency.

## Solution

Re-implement file-based ledger storage, with a focus on writing as fast as possible, as well as providing more specifically useful API and other characteristics.

## Goals

- Remove rocksdb dependency
- Store ledger data reliably
- Exceed current DbLedger performance

## Design

There are two primary components: `BlobStore` and `KVStore`

### `BlobStore`

Essentially a struct that should behave as close to `DbLedger` as possible so that replacement is easy. It relies on the `KVStore` for all actual storage

### `KVStore`

Key-value persistence layer that maps keys to a value.
`Value` is just a type alias for clarity: `type Value = Vec<u8>`. All operations must be ACID transactions to whatever degree possible.

Operations:

- `put (key: Key, value: Value) -> Result<()>`
- `get (key: Key) -> Result<Option<Value>>`
- `delete (key: Key) -> Result<bool>`
  - return value indicates if there was something to delete
- `put_many (rows: [(K, V)]) -> Result<()>`
- `get_many (keys: [Key]) -> Result<Vec<Value>>`
- `delete_many (keys: [Key]) -> Result<()>`
- `scan(range: RangeInclusive<Key>) -> Result<impl Iterator<Item=(Key, Value)>>`

The storage strategy is essentially a simple log-structured merge tree (LSMT) approach. There is an in-memory table and series of on-disk `SSTables` [1]. When storing a record, it is first written to a write-log, then added to the in-memory table.

When this table fills up, it is then dumped to disk in two files. 1 file simple contains the `(Key, Value)` pairs. The second file is the index file, which contains some metadata, and a list of `(Key, Loc)` pairs where `Loc` represents to location and extent of the record in the data file. These files combined form a logical `SSTable`. When the in-memory table is dumped, it creates an `SSTable` of level 0. As level 0 fills up, the compaction process begins.

The compaction strategy is known as "levelled compaction" and is mostly a simplified version of what's described by ScyllaDB and LevelDB [2][3].

## Implementation

Performance is very critical for this feature. I have rudimentary benchmarks that indicate same-or-better perf when compared to the current `DbLedger` but not very rigorous or well-documented ones. I will create them.

### Questions

- Should most of this comment (probably everything above this) be expanded on and moved to something like a proposal in the book?
- Should this be a new crate in the workspace?
  - Currently everything is mostly a new module in the solana crate. This seems like it could be a separate crate. That would make boundaries clear and might make builds, test, and benchmarking simpler (and possibly faster).

### In-flight

- [x] principled indexing and storage using simple LSMTree approach
  - [x] insert
  - [x] flushing in-memory table
  - [x] merging on disk tables
  - [x] read
    - [x] simple read
    - [x] merge read
  - [x] range scan
  - [x] recovery
    - [x] in-memory write buffer
    - [x] SSTable indexes
  - [ ] integrate
- [ ] Change naming to be more inline with recent developments (block/slot/etc)
- [x] Testing
  - [x] basic integration tests
  - [ ] utility functions for testing
  - [ ] Isolate and test unit-testable components or find some way to do I/O free testing
- [ ] Benchmarks
  - [x] basic benchmarks
- [ ] Documentation
  - [ ] document all modules, types, and methods exposed to the rest of the crate
  - [ ] document important internal modules, types, and methods

1. <https://www.igvita.com/2012/02/06/sstable-and-log-structured-storage-leveldb/>
2. <https://github.com/scylladb/scylla/wiki/SSTable-compaction-and-compaction-strategies#leveled-compaction>
3. <https://github.com/google/leveldb/blob/master/doc/impl.md>

### PR DESCRIPTION
**Pull request by MarkJr94**
This is largely a WIP PR for work on this issue https://github.com/solana-labs/solana/issues/2566

Everything about `BlobStore` is out-dated, especially due to `BlockTree`.

A fuller design description will be posted as things progress. For now please see the tracking issue.

#### Problem
rocks
#### Summary of Changes
simple key-value store to persist ledger data

Fixes #2566

### CODE REVIEW

**Comment by garious** [comment]
Hey @MarkJr94, anything you can do to get us off RocksDB sooner than later would be greatly appreciated!

**Comment by garious** [comment]
No merge commits please! Just rebase!

**Comment by aeyakovenko** [comment]
Can you rebase and squash this pr, it seems like it’s pulling in a lot of commits that are in master.

**Comment by MarkJr94** [comment]
@aeyakovenko yes before i change this from draft status today I'll fix up the commit history. worst case i'll just force-push from a cleaner branch where i have copied over changes manually, which would be fairly easy to do since most changes don't affect anything els.e

File: core/src/blocktree.rs [hunk 1]
```diff
@@ -3,121 +3,80 @@
...
 use bincode::{deserialize, serialize};
-use byteorder::{BigEndian, ByteOrder, ReadBytesExt};
+
 use hashbrown::HashMap;
-use rocksdb::{
-    ColumnFamily, ColumnFamilyDescriptor, DBRawIterator, IteratorMode, Options, WriteBatch, DB,
-};
-use serde::de::DeserializeOwned;
+
+use rocksdb;
```
Comment (garious) [inline core/src/blocktree.rs]: Should this have a: `#[cfg(not(feature = "kvstore"))]`

Comment (MarkJr94) [reply core/src/blocktree.rs]: Yes, fixing

**Comment by mvines** [comment]
yikes, merging while CI is still running.  📿

**Comment by garious** [comment]
NBD, he only added a cfg() macro that only affected a configuration uncovered by CI.

**Comment by mvines** [comment]
Maybe :)

**Pull request merged**